### ISSUE TITLE
provide binaries for Windows, MacOS, Android and Linux

### ISSUE THREAD

**Issue by mk270** [mk270/whitakers-words#2]

**Comment by mscuthbert** [comment]
I have a Mac binary based on the older source repository online at http://www.trecento.com/misc/Whitakers_Words.zip -- I don't know if it's something that would want to be hosted here until a binary based on the newest version can be added, but feel free to use it as a resource w/ a halfway decent icon for a Mac version.  The only Mac binaries I could find were compiled for PowerPC and don't run on Lion or above.  I can also provide the icon separately (it's based on a manuscript from the Library of Congress which does not claim copyright on photographs of medieval manuscripts, so it should be copyright cleared.  I altered the initial "V" to a "W" for this program).

**Comment by mk270** [comment]
Michael,

thanks for chipping in; your concentration on the copyright status of the artefacts is exactly right - there are a couple of IPR pitfalls we need to avoid, particularly the fact that Ada executables have to be compiled with the right compiler - a very popular version of the GNAT compiler links executables against a GPL-ed version of the runtime, which obligates distributors to relicense their code under the GPL, which we can't do without permission from Whitaker's estate.

I think the thing to do is create a landing page associated with this preservation effort, with links to existing binaries, and a note to the effect that we'll distribute binaries from the current source code in due course. So your link is useful - maybe we should add it to the README page for the time being, and worry about the wider issues later.

**Comment by mscuthbert** [comment]
Ah! i'm mostly a Python hack (attempted to rewrite Words in Python once, but didn't get very far) so I didn't think of the linking problems (yes, my binary was built with GNAT, so it's probably not appropriate for inclusion, but the icon is problem free).

I think that the copyright on the project as provided by WW is fine, but do the modifications to the source need to be specified as bound by the same license? I'd hate to have the project derailed at some point because some contributor claims exclusive copyright on one of their contributions.  Linking to the binary where it is is fine.  My site has been stable for at least 15 years and hosted at MIT so I'm sure it'll be fine until we can get up-to-date binaries here.

**Comment by mk270** [comment]
The main complication with GNAT is just _names_: there are three Ada compilers with confusingly similar names (FSF GNAT, GNAT GPL, GNAT Pro); it's not hard to distinguish them once you know the names, and it's easy to confuse them until you do. We need to use FSF GNAT for redistributable binaries; this is just a question of getting the build environment set up properly - we can stick in a check for the right version of the compiler at the appropriate point; I've made ticket #12 for this

Re contributions: yes, we're going to have to have a contributor agreement thing to avoid the issue you refer to - luckily the issue is currently negligible and we know about it.

**Comment by mk270** [comment]
Please re-file your wishlist/todo items re hotlinking (etc) as separate issues here on github

**Comment by mk270** [comment]
@mscuthbert , are you willing and able to build a new MacOS binary when the time arises? Also, where did the additional material in the ZIP file you host come from?

**Comment by mscuthbert** [comment]
Definitely.  I'll figure out how to make it be a standalone Terminal application by then (right now, when you exit the program it just dumps you to the command line).  The only additional material that should be there is the icon.  I'm not sure what else is there.

**Comment by Fuco1** [comment]
I'm not sure if travis-ci can be configured for different systems than debian (or whatever it runs), but it would be worth to check it out. A CI server which would automatically do builds would make things much simpler.

**Comment by mk270** [comment]
@Fuco1  - we don't actually have any test cases for Travis to run, though?

**Comment by mk270** [comment]
@mscuthbert cheers! I think dumping on the commandline is fine for v1.98 - making it behave as a typical Windows and MacOS application is a little further in the future :) .. but don't let that hold you back.

**Comment by Fuco1** [comment]
@mk270 I ment that it would run builds instead of tests.  I don't know if it can do that with the free version though...

**Comment by mk270** [comment]
Ok, good point; separately we need someone who knows how to package commandline apps for MacOS to step forward and write us a Makefile target.

**Comment by mk270** [comment]
@Fuco1 a trivial test case has now been added, and runs via Travis.

I am hoping to get some funds to get an intern to compile a rigorous test suite (that is, someone who knows Latin and has some sense of programming).

**Comment by mk270** [comment]
@mscuthbert MacOS help would certainly be appreciated now - I have also made a web version, which I'm sure many people will be happy to use

**Comment by mscuthbert** [comment]
@mk270 Sorry that I missed this message.  I'll try to get to it this week. The file I compiled from the old 1.97 version could definitely go up unless there are license issues with the compiler I used -- the current version we should mark as not compatible with Intel macs running OS X Lion or above (i.e., 99.9% of Macs out there). Thanks all!

**Comment by mk270** [comment]
@mscuthbert no worries! Thank you for following up again.

I reckon we can infer the compiler used with:

```
$ strings bin/words | grep "GNAT Version"
```

For FSF GNAT, I get `GNAT Version: 4.6`; for GNAT GPL, I get `GNAT Version: GPL 2015 (20150428-49)`

Any old copies you can find, please check for whether they exhibit the bug in issue #46 - I'm still unsure if this is a config problem, a bug introduced by Whitaker, or a bug introduced by us.

Do you have a GNAT environment for MacOS available?

**Comment by mscuthbert** [comment]
:-( I have the GPL version.  Just spent an hour trying to compile a non-GPL version but I haven't been able to make any progress (I'm not an Ada hacker, I'm afraid).  If someone w/ a Mac can figure out a compilation solution, that'd be very helpful.  The compiled words/ directory can then be copied (wholesale or selectively if you know what you're doing) into the Whitakers Words.app directory under Contents/Resources/ and the app will just use the newer words binary (it's not very smart; it'll execute anything called "words" in that folder).

**Comment by mscuthbert** [comment]
It does have the Issue #46 bug, I'm afraid. (commented there).

**Comment by mk270** [comment]
@mscuthbert thanks for this report - that you've got this version compiling at all on MacOS is an important result. It may be that we should do #29 and sacrifice the advantages of gprbuild for portability, which I think would make a non-GNAT GPL version work easily on MacOS - but that doesn't need to be decided yet.

**Comment by adunning** [comment]
It would be really great to have this added to Homebrew for OS X.

**Comment by mk270** [comment]
@adunning  Good point! I had forgotten about Homebrew, not being a MacOSer. Is this something you'd be able to help with, or is this just something you are suggesting we do?

**Comment by adunning** [comment]
I thought I could, as I've done this with a few other packages, and it's normally quite straightforward – but the build dependencies also need to be packaged, and I hadn't realized that nobody has done FSF GNAT. I'm sure it's possible, but I don't know enough about Ada to take care of it at the moment.

By the way, if you're not already aware of it, there's also [Interpres](https://sites.google.com/site/erikandremendoza/) for OS X.

**Comment by mk270** [comment]
Ok, thanks - it is becoming clear that the availability of FSF GNAT on MacOS and Windows is a _strategic_ problem for WORDS.

I'm aware of Interpres. The general thrust of splitting WORDS into libraries is to allow these people making wrappers to get typed information about inflections, rather than the string blob that's currently available. Interpres does seem to use FSF GNAT, albeit an ancient version.

**Comment by adunning** [comment]
I'm afraid so. I almost wonder if it would be easier to see whether the author of [Parsley](http://parsley.goldibex.com) would be open to having someone pick it up and expand it with the Words dictionary.

**Comment by mk270** [comment]
It looks like gprbuild is now available on github, which may be a game-changer.

**Comment by air-wreck** [comment]
I don't know if any help is still needed on the Linux side of this, but I've built on Debian with FSF GNAT and packaged it into [an AppImage](https://air-wreck.gitlab.io/files/misc/words.AppImage) (along with the `libgnat-6` runtime). It seems to work on Debian, Ubuntu, and Fedora. I'm new to packaging things, so I'm not sure how widely it will work (e.g. I suspect it won't work on 32-bit platforms), but I'd be willing to look into it more if help is still needed with binaries on Linux.

**Comment by ids1024** [comment]
> e.g. I suspect it won't work on 32-bit platforms

Presumably, though it shouldn't be hard to build a 32-bit one as well. I haven't installed an AppImage, let alone built one, but presumably that uses some kind of build script. Could you post that somewhere?

Lately I've been working on a docker image for [an FSF GNAT toolchain targeting Android](https://github.com/ids1024/ada-android-docker) for [my words Android app](https://github.com/ids1024/whitakers-words-android), which now has words (Whitaker's last release, currently, not this repo) building on Travis CI. Of course, it's much easier when you're targeting normal Linux instead of Android; you don't need to build a custom cross toolchain.

**Comment by air-wreck** [comment]
There's not much of a script; I just built the binary with FSF GNAT, followed [these instructions](https://github.com/AppImage/docs.appimage.org/blob/master/source/packaging-guide/manual.rst), and then ran [their tool](https://github.com/AppImage/AppImageKit). Besides patching a path and copying in the runtime, everything worked very smoothly. On the other hand, it should be fairly easy to automate this process.

**Comment by Alex-Vasile** [comment]
Hi all, I managed to build a Mac binary based on the newest version of the source from here. I'd love to contribute it, and help out. But I am not sure if i'm using the "correct" version of GNAT, or what to help with in general.

When you all refer to the FSF GNAT I assume you mean the one from here (https://www.gnu.org/software/gnat/). I ended up using GNAT_GCC from here (https://sourceforge.net/projects/gnuada/files/GNAT_GCC%20Mac%20OS%20X/10.1.0/native/) since it was straight forward to make. To be honest I can't tell the difference. The license files in both of them look the same to me.

When I run `gnatmake --version` I get: `GNATMAKE 10.1.0`

Alternatively if this is not the correct verion of gnat, I would be willing to give installing the correct one a go, if someone can point me in the right direction.

**Comment by ids1024** [comment]
> I ended up using GNAT_GCC from here

It looks to me like that Sourceforge project is the FSF version of GNAT, but packaged by some third party (though I'm not entirely sure)? Then I guess that should be fine.

It's the "community" edition of AdaCore's Ada that's potentially a problem. Though it depends whether the informal license the original words code was placed under counts as "GPL compatible".

> I'd love to contribute it, and help out. But I am not sure if i'm using the "correct" version of GNAT, or what to help with in general.

I'm not sure what @mk270's thoughts are, but it seems to me the best way to offer binaries would probably be to add GitHub actions workflows building for Windows, Mac, and Linux.

Not sure what the best way to use GNAT on Github actions would be, since Homebrew doesn't seem to package it. On Linux it's generally packaged and isn't an issue.

**Comment by Alex-Vasile** [comment]
> Not sure what the best way to use GNAT on Github actions would be, since Homebrew doesn't seem to package it. On Linux it's generally packaged and isn't an issue.

There's no Homebrew, but the Sourceforge install method was fairly pain-free. Do you know how github's actions handle unsigned code on the Mac instances? Because the unsigned nature of Sourceforge approach meant I had to manually allow each executable in GNAT to run.

**Comment by mk270** [comment]
@ids1024 Yes - let's use GitHub Actions as you suggest

I understand there's a move afoot within the Ada community to change the basis for licensing various versions of the compiler. This potentially has some impact on how we'd facilitate building WORDS from source. It's a concern I've had for some years, but I don't yet know enough about it to know whether the proposed changes are net positive or negative for our situation here.
